### PR TITLE
Add planning based scan hints 

### LINKED ISSUE #2542: Add rules for applying scan hints in query planning
To take better advantage of https://accumulo.apache.org/docs/2.x/administration/scan-executors allow rule based configuration of hints to be added during query planning.

### PR DESCRIPTION
**Pull request by FineAndDandy**
closes #2542

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/config/ScanHintRule.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+     * Get the table the hint is applied to when apply()=true
+     *
+     * @return the table the hint
+     */
+    String getTable();
+
+    /**
+     * The hint name to use when apply()=true
+     *
+     * @return
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/config/ScanHintRule.java]: java doc?

File: warehouse/query-core/src/main/java/datawave/query/config/ScanHintRule.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    String getHintName();
+
+    /**
+     * The hint value to use when apply()=true
+     *
+     * @return
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/config/ScanHintRule.java]: java doc?

File: warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java [hunk 3]
```diff
@@ -1867,6 +1868,14 @@ public static boolean validateJunctionChildren(JexlNode node, boolean failHard)
         return valid;
     }
 
+    public static NodeTypeCount getIvarators(JexlNode node) {
+        return NodeTypeCountVisitor.countNodes(node, QueryPropertyMarker.MarkerType.EXCEEDED_VALUE, QueryPropertyMarker.MarkerType.EXCEEDED_OR);
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java]: See QueryPropertyMarker.IVARATOR_TYPES, in addition to these two types the EXCEEDED_TERM is also an ivarator type.

File: warehouse/query-core/src/main/java/datawave/query/planner/scanhints/ExecutorScanHintRule.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package datawave.query.planner.scanhints;
+
+import org.apache.commons.jexl3.parser.JexlNode;
+
+import datawave.query.config.ScanHintRule;
+
+/**
+ * Scan hints that are used to select an executor in accumulo
+ *
+ * @see <a href="https://accumulo.apache.org/docs/2.x/administration/scan-executors">https://accumulo.apache.org/docs/2.x/administration/scan-executors</a>
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/planner/scanhints/ExecutorScanHintRule.java:10]: linking to docs, nice

File: warehouse/query-core/src/main/java/datawave/query/planner/scanhints/IvaratorScanHint.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+public class IvaratorScanHint extends ExecutorScanHintRule {
+    @Override
+    public boolean isChainable() {
+        return true;
+    }
+
+    @Override
+    public String getTable() {
+        return "shard";
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/planner/scanhints/IvaratorScanHint.java]: This table name should be configurable. While "shard" is the default shard table name (see the TableName constants), there's nothing stopping a user from creating a shard table with the name like "datawave.shard" or "test.shard". In those cases scan hints wouldn't be applied.

I don't consider this a blocking issue, feel free to do this later.

File: warehouse/query-core/src/main/java/datawave/query/planner/scanhints/IvaratorScanHint.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.apache.commons.jexl3.parser.JexlNode;
+
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.jexl.NodeTypeCount;
+
+/**
+ * Detect ivarators and set the executor scan hint to ivarator
+ */
+public class IvaratorScanHint extends ExecutorScanHintRule {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/planner/scanhints/IvaratorScanHint.java:11]: Could you add a test that runs through several variations of queries to prove some positive and negative cases? It doesn't need to be as expansive as the RewriteNegationsVisitorTest, but that gives you an idea of what I'm looking for.

**Review by apmoriarty — CHANGES_REQUESTED** [review]

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 7]
```diff
@@ -683,10 +683,8 @@ public static void validateQuerySize(String lastOperation, final JexlNode queryT
 
         // now check whether we are over the ivarator limit
         if (maxIvaratorThreshold >= 0) {
-            NodeTypeCount nodeCount = NodeTypeCountVisitor.countNodes(queryTree, QueryPropertyMarker.MarkerType.EXCEEDED_VALUE,
-                            QueryPropertyMarker.MarkerType.EXCEEDED_OR);
-            int totalIvarators = nodeCount.getTotal(QueryPropertyMarker.MarkerType.EXCEEDED_VALUE)
-                            + nodeCount.getTotal(QueryPropertyMarker.MarkerType.EXCEEDED_OR);
+            NodeTypeCount nodeCount = JexlASTHelper.getIvarators(queryTree);
+            int totalIvarators = JexlASTHelper.getIvaratorCount(nodeCount);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:688]: nice

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 8]
```diff
@@ -1111,6 +1111,41 @@ protected boolean timedCheckForSortedUids(QueryStopwatch timers, String stage, S
...
     }
 
+    protected void timeScanHintRules(QueryStopwatch timers, String stage, ShardQueryConfiguration config) {
+        TraceStopwatch stopwatch = timers.newStartedStopwatch("DefaultQueryPlanner - " + stage);
+
+        // apply runtime scan hints
+        if (config.isUseQueryTreeScanHintRules()) {
+            for (ScanHintRule<JexlNode> hintRule : config.getQueryTreeScanHintRules()) {
+                if (hintRule.getTable() != null && hintRule.apply(config.getQueryTree())) {
+                    Map<String,String> tableHints = config.getTableHints().get(hintRule.getTable());
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: My only thought here is we are assuming that config.getTableHints is not returning a copy.  Maybe we create a local map and the config.setTableHints at the end.

File: warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java [hunk 9]
```diff
@@ -3156,4 +3167,20 @@ protected Object readResolve() throws ObjectStreamException {
         this.fstCount = new AtomicInteger(0);
         return this;
     }
+
+    public boolean isUseQueryTreeScanHintRules() {
+        return useQueryTreeScanHintRules;
+    }
+
+    public void setUseQueryTreeScanHintRules(boolean useQueryTreeScanHintRules) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java:3208]: We need to add delegate methods to the ShardQueryLogic so that we can actually configure these.

File: warehouse/query-core/src/main/java/datawave/query/planner/scanhints/PriorityScanHintRule.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import org.apache.commons.jexl3.parser.JexlNode;
+
+import datawave.query.config.ScanHintRule;
+
+/**
+ * Scan hints that are used to set a priority in accumulo,
+ *
+ * @see <a href="https://accumulo.apache.org/docs/2.x/administration/scan-executors">https://accumulo.apache.org/docs/2.x/administration/scan-executors</a>
+ */
+public abstract class PriorityScanHintRule extends BaseScanHintRule<JexlNode> {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/planner/scanhints/PriorityScanHintRule.java:12]: What are we thinking of for an implementation of this one?

**Review by alerman — DISMISSED** [review]
LGTM

File: warehouse/query-core/src/main/java/datawave/query/planner/scanhints/PriorityScanHintRule.java (hunk 10 above)
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/planner/scanhints/PriorityScanHintRule.java:12]: I'd like to put together another PR that leverages the HintScanPrioritizer (from https://accumulo.apache.org/docs/2.x/administration/scan-executors) and has a weighted visitor to arrive at a priority based on the number of and type of query terms.

File: warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java (hunk 9 above)
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java:3208]: now how long its been since I had a query update is showing....

**Review by ivakegg — APPROVED** [review]

**Review by apmoriarty — APPROVED** [review]
